### PR TITLE
User holdings using SuiClient Api

### PR DESCRIPTION
**Pull request by OctoSauce**

### CODE REVIEW

File: src/getUserHoldingsUsd.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  } catch (error) {
+    console.error(`Error fetching price of ${pool}:`, error);
+    return undefined;
+  }
+}
+
+function mergeDuplicateHoldings(
+  userTokens: [string, string, string][],
+): [string, string, string][] {
+  const map = new Map<string, number>();
```
Comment (jangid) [inline src/getUserHoldingsUsd.ts:184]: give some name to map. what it is used for, purpose etc.

File: src/getUserHoldings.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      conversionMap.get(pool as PoolName) as string,
+    );
+    const tokens = new Decimal(xTokens).mul(conversion).toFixed(4).toString();
+    return [owner, pool, tokens];
+  });
+
+  return userTokens;
+}
+
+// function mergeDuplicateHoldings(
```
Comment (jangid) [inline src/getUserHoldings.ts:164]: remove commented lines

File: src/getUserHoldings.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
+import Decimal from "decimal.js";
+import { getPoolConversionMap, poolIdPoolNameMap } from "./common/maps";
+import { getHolders } from "./getHolders";
+import { getReceipts } from "./utils/getReceipts";
+
+import { SuiObjectResponse } from "@mysten/sui/client";
+import { PoolName } from "./common/types";
+
+export async function getUserTokens(params?: {
+  pools?: string[];
```
Comment (jangid) [inline src/getUserHoldings.ts:10]: change to poolName and create a type for complex params.

File: src/getUserHoldings.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  }
+  const receipts = await getReceipts({
+    pools: params?.pools,
+    owners: owners,
+  });
+  const userTokens = parseTokensFromReceipts(receipts);
+  return userTokens;
+}
+
+type alphaReceiptFields = {
```
Comment (jangid) [inline src/getUserHoldings.ts:29]: move this to "./src/types.ts" and Type names should be CapilatlizedCamelCase.

File: src/getUserHoldings.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        },
+      ];
+    };
+  };
+  pool_id: string;
+  unlocked_xtokens: string;
+  xTokenBalance: string;
+};
+
+type otherRecceiptFields = {
```
Comment (jangid) [inline src/getUserHoldings.ts:88]: same as previous comment.

**Review by jangid — CHANGES_REQUESTED** [review]

File: src/common/maps.ts [hunk 6]
```diff
@@ -429,3 +439,98 @@ export const poolIdQueryInvestorMap: { [key: string]: string } = {
   "0x045e4e3ccd383bedeb8fda54c39a7a1b1a6ed6a9f66aec4998984373558f96a0":
     "navxSuiInvestor",
 };
+
+// Pagination needed for more than 50 pools
+export async function getPoolConversionMap(): Promise<Map<PoolName, string>> {
```
Comment (jangid) [inline src/common/maps.ts:444]: rename to getPoolExchangeRateMap

File: src/common/maps.ts [hunk 7]
```diff
@@ -429,3 +439,98 @@ export const poolIdQueryInvestorMap: { [key: string]: string } = {
...
+    );
+    const sqrtPrice = poolDetails.content.fields.current_sqrt_price;
+    poolNameToSqrtPriceMap.set(pool as PoolName, sqrtPrice);
+  }
+
+  return poolNameToSqrtPriceMap;
+}
+
+// Pagination needed for more than 50 pools
+export async function getCetusInvestorTicksMap(): Promise<{
```
Comment (jangid) [inline src/common/maps.ts:493]: return type object should have key of type PoolName and not string.

**Review by jangid — CHANGES_REQUESTED** [review]

File: src/sui-sdk/transactions/fetchTransactions.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        showInput: true,
+        showObjectChanges: true,
+      },
+    });
+    if (res.data.length !== 0) {
+      const lastTx = res.data[res.data.length - 1];
+      const firstTx = res.data[0];
+      const laterTime = firstTx.timestampMs as string;
+      const earlierTime = lastTx.timestampMs as string;
+      if (Number(laterTime) < params.startTime) {
```
Comment (jangid) [inline src/sui-sdk/transactions/fetchTransactions.ts:26]: why we need this check? Are the start and end date params not working in queryTransactionBlocks?

File: src/sui-sdk/transactions/types.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { TransactionFilter } from "@mysten/sui/client"
+
+export type FetchTransactionParams = {
```
Comment (jangid) [inline src/sui-sdk/transactions/types.ts:3]: Add 'order' attribute to the FetchTransactionParams. So that we can reuse it for specific order.

**Review by jangid — CHANGES_REQUESTED** [review]

File: src/getDataFromEvents.ts
Comment (jangid) [inline src/getDataFromEvents.ts:1]: Rename this file to getActivity.ts

**Review by jangid — CHANGES_REQUESTED** [review]

File: src/getHolders.ts
Comment (jangid) [inline src/getHolders.ts:1]: How is `getHolders` function different from `getHoldersFromEvents` in the file `getDataFromEvents`? Aren't both of them returning same accounts in the same time-span?

**Review by jangid — CHANGES_REQUESTED** [review]

File: src/sui-sdk/transactions/fetchTransactions.ts (hunk 8 above)
Comment (OctoSauce) [reply src/sui-sdk/transactions/fetchTransactions.ts:26]: queryTransactionBlocks is a suiClient API, it doesn't have timestamp checks. This is for having timestamp check.

**Comment by OctoSauce** [comment]
Resolve overlap with timestamp checks in src/sui-sdk/transactions/fetchTransactions.ts The transactions should be exactly in the timestamp range, not just based on pages

File: src/getHolders.ts
Comment (OctoSauce) [reply src/getHolders.ts:1]: Yes both use different ways

**Comment by OctoSauce** [comment]
> Resolve overlap with timestamp checks in src/sui-sdk/transactions/fetchTransactions.ts The transactions should be exactly in the timestamp range, not just based on pages

Resolved

**Pull request closed without merging**